src/rust-getting-started/string-tutorial.md

```diff
@@ -0,0 +1,238 @@
+  - `HSTRING`: A handle to a Windows Runtime string. HSTRINGS are UTF-16 and immutable. 
+
+Note that you can pass BSTR or HSTRING to functions expecting a PCWSTR. 
+
+Unfortunately, none of these types match Rust types one-to-one. However, we can use the `windows-strings` crate to help us. 
+
+## Types of API Functions (Narrow or Wide)
+
+The Win32 API divides string functions into their "narrow" version (ending in 'A', like `MessageBoxA`), and their "wide" version (ending in 'W', like `MessageBoxW`). Narrow versions of the API expect u8 byte strings, encoded using the current thread's code page, while wide versions expect UTF-16. 
+
+As a general recommendation, you should prefer wide versions; it's much easier to convert between Rust's UTF-8 strings and Windows' UTF-16.
+
+## Calling APIs that Consume Strings
+
+Let's look at an example using the simple [`MessageBox` function](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/UI/WindowsAndMessaging/fn.MessageBoxW.html), which displays a pop up dialog box. We'll use the wide version (`MessageBoxW`) for this example. 
+
+If you want to call a Windows API using a string literal, the `windows-strings` crate has macros to generate windows strings from string literals:
+  - `h!` generates an HSTRING from a string literal, adding a null terminator and converting to UTF-16.
+  - `w!` does the same, but generates a PCWSTR instead of an HSTRING. 
+  - `s!` generates a PCSTR with a null terminator. **TODO: does this do any conversion**
+
+If we wanted to call the message box we could use the windows crate with the `Win32_UI_WindowsAndMessaging` feature, and call:
+
+```rust
+fn show_message_box_literals() -> MESSAGEBOX_RESULT {
+    // use string literals when calling a message box. 
+    let text = h!("Hello from rust!");
+    let caption = h!("From Rust");
+
+    unsafe {
+        UI::WindowsAndMessaging::MessageBoxW(None, 
+            text, 
+            caption,
+            UI::WindowsAndMessaging::MESSAGEBOX_STYLE(0) // message box OK
+        )
+    }
+}
+```
+
+This works, but what if we wanted to call the same function with a Rust string? That gets slightly more complicated. We could manually convert to a UTF-16 series of bytes, and add the null terminator ourselves, like this:
+
+```rust
+fn show_message_box_u16_conversion(text: &str, caption: &str) -> MESSAGEBOX_RESULT {
+    // convert our text and caption to UTF-16 bytes
+    let mut text = text.encode_utf16().collect::<Vec<u16>>();
+    let mut caption = caption.encode_utf16().collect::<Vec<u16>>();
+
+    // add null terminators
+    text.push(0u16);
+    caption.push(0u16);
```

Comment:
In defense of the Standard Library, this is typically simpler using encode/chain/collect.